scipy/sparse/csgraph/_flow.pyx

```diff
@@ -0,0 +1,398 @@
+    residual_matrix = csr_matrix((residual_array, m.indices, m.indptr),
+                                 shape=m.shape)
+    source_flow = residual_array[m.indptr[source]:m.indptr[source + 1]]
+    return MaximumFlowResult(source_flow.sum(), residual_matrix)
+
+
+def _add_reverse_edges(a):
+    """Add reversed edges to all edges in a graph.
+
+    This adds to a given directed weighted graph all edges in the reverse
+    direction and give them weight 0, unless they already exist.
+
+    Parameters
+    ----------
+    a : csr_matrix
+        The square matrix in CSR format representing a directed graph
+
+    Returns
+    -------
+    res : csr_matrix
+        A new matrix in CSR format in which the missing edges are represented
+        by explicit zeros.
+
+    """
+    # This is equivalent to just looping over all (i, j) and letting
+    # the weight of (j, i) be zero if it is not already defined.
+    # Using NumPy to do so is several orders of magnitude faster
+    # than the naive solution. We first create create the sorted
+    # list {i*n + j} for nonzero (i, j), then use binary search
+    # to determine if j*n + i belongs to the list. To ensure that
+    # the haystack below is sorted, we presort the column indices if
+    # necessary.
+    if not a.has_sorted_indices:
+        a = a.sorted_indices()
+    acoo = a.tocoo()
+    rows = acoo.row
+    cols = acoo.col
+    n = a.shape[0]
+    haystack = rows*n + cols
+    needles = cols*n + rows
+    indices = np.searchsorted(haystack, needles)
+    # As searchsorted returns a.nnz when a given needle is
+    # larger than the largest element of the haystack, we
+    # explicitly add an element not in needles at the end.
+    haystack = np.concatenate([haystack, [-1]])
+    indexed = haystack[indices]
+    diff = indexed != needles
+    # At this point, diff contains the indices of the (j, i)
+    # to be added. We make use of the COO initializer for CSR
+    # matrices which ensures that 0s are added explicitly.
+    indices_to_add = np.where(diff)[0]
+    data_to_add = np.zeros((indices_to_add.shape[0],), dtype=ITYPE)
+    acoo.row = np.concatenate([acoo.row, cols[indices_to_add]])
+    acoo.col = np.concatenate([acoo.col, rows[indices_to_add]])
+    acoo.data = np.concatenate([acoo.data, data_to_add])
+    return acoo.tocsr()
+
+
+def _make_edge_pointers(a):
+    """Create for each edge pointers to its reverse and its tail."""
+    # Just as above, one way to match (i, j) with (j, i) is to match
+    # i*n + j with j*n + i, which we can do by ensuring that we
+    # have a sorted collection of {i*n + j} and using binary search
+    # to find j*n + i here.
+    n = a.shape[0]
+    acoo = a.tocoo()
+    rows = acoo.row
+    cols = acoo.col
+    haystack = rows*n + cols
+    needles = cols*n + rows
+    rev_edge_ptr = np.searchsorted(haystack, needles).astype(ITYPE)
+    return rev_edge_ptr, rows
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef int[:] _edmonds_karp(
```

Comment:
oops sorry my previous comment was wrong. All of these type declarations should be `ITYPE_t` rather than `int`.

Reply:
But still retaining `ITYPE` where we had `np.int32`, right? If so, [`c153318`](https://github.com/scipy/scipy/pull/10566/commits/c153318aae5fa6832b1713acd90fc186a7d56169) should take care of that.

Reply:
yes correct. using `ITYPE` where you'd use a numpy dtype, and `ITYPE_t` for Cython memoryviews or Cython `ndarray` objects (you don't use the latter, but other `csgraph` functions do).

Reply:
Bonus question: This doesn't harm performance, right?

Reply:
indeed, identical performance-wise